schematic/store/synapse.py

```diff
@@ -50,8 +52,100 @@
+    """
+    syn: an object of type synapseclient.
+    manifest_id: id of a manifest  
+    """
+    syn: synapseclient.Synapse
+    manifest_id: str
+
+    def _download_manifest_to_folder(self):
+        """
+        try downloading a manifest to local cache or a given folder
+        manifest
+        Return: 
+            manifest_data: A new Synapse Entity object of the appropriate type
+        """
+         # TO DO: potentially deprecate the if else statement because "manifest_folder" key always exist in config
+        if CONFIG["synapse"]["manifest_folder"]:
+            manifest_data = self.syn.get(
+                    self.manifest_id,
+                    downloadLocation=CONFIG["synapse"]["manifest_folder"],
+                    ifcollision="overwrite.local",
+                )
+        else:
+            manifest_data = self.syn.get(
+                        self.manifest_id,
+                    )
+        return manifest_data 
+
+    def _entity_type_checking(self):
+        """
+        check the entity type of the id that needs to be downloaded
+        Return: 
+             if the entity type is wrong, raise an error
+        """
+        # check the type of entity
+        entity_type = entity_type_mapping(self.syn, self.manifest_id)
+        if entity_type  != "file":
+            logger.error(f'You are using entity type: {entity_type}. Please provide a file ID')
+
+    @staticmethod
+    def download_manifest(self, newManifestName: str="", manifest_df: pd.DataFrame=pd.DataFrame()) -> Union[str,File]:
+        """
+        Download a manifest based on a given manifest id. 
+        Args:
+            newManifestName(optional): new name of a manifest that gets downloaded.
+            manifest_df(optional): a dataframe containing name and id of manifests in a given asset view
+        Return: 
+            manifest_data: synapse entity file object
+        """
+
+        # enables retrying if user does not have access to uncensored manifest
+        # pass synID to synapseclient.Synapse.get() method to download (and overwrite) file to a location
+        manifest_data = ""
+
+        # check entity type
+        self._entity_type_checking()
+
+        # download a manifest
+        try:
+            manifest_data = self._download_manifest_to_folder()
+        except(SynapseUnmetAccessRestrictions, SynapseAuthenticationError):
+            # if there's an error getting an uncensored manifest, try getting the censored manifest
+            if not manifest_df.empty:
+                censored_regex=re.compile('.*censored.*')
+                censored = manifest_df['name'].str.contains(censored_regex)
+                new_manifest_id=manifest_df[censored]["id"][0]
+                self.manifest_id = new_manifest_id
+                try: 
+                    manifest_data = self._download_manifest_to_folder()
+                except (SynapseUnmetAccessRestrictions, SynapseAuthenticationError) as e:
```

Comment:
Do these errors also need to be caught?

Reply:
hmm my logic here is: 
in `getDatasetManifest` function, first [`_get_manifest_id`](https://github.com/Sage-Bionetworks/schematic/blob/develop-add-uWSGI-two/schematic/store/synapse.py#L463) function would first try to figure out what manifest that we wants to download. 
* If censored and uncensored manifests are both available, we try downloading the uncensored manifest first. 
* If there's an error downloading the uncensored manifest because of access restricted (that's why I have the first try and except block [here](https://github.com/Sage-Bionetworks/schematic/blob/develop-add-uWSGI-two/schematic/store/synapse.py#L117)), try downloading the uncensored manifest first
* if there's still an error downloading the censored manifest, the error of not having access gets raised again, which is why I have the code here. 

I think the general logic is what we have in the old code. If I don't catch the errors here, those errors might be raised by python client. But I think here maybe I could emphasize that users don't have access to both censored and uncensored manifest? So something like: 
```
                except (SynapseUnmetAccessRestrictions, SynapseAuthenticationError) as e:
                    raise PermissionError("You don't have access to censored and uncensored manifests in this dataset.") from e
```

Reply:
Ahh I see, that makes sense. Thanks for explaining!